lib/ditty/controllers/main.rb

```diff
@@ -145,21 +146,47 @@ def redirect_path
       if env['omniauth.auth']
         # Successful Login
         user = User.find(email: env['omniauth.auth']['info']['email'])
-        #Check Identity for Password Expiry
+        # Check Identity for Password Expiry
         password_expiry_check(user)
         self.current_user = user
-        broadcast(:user_login, target: self, details: "IP: #{request.ip}")
-        halt 200 if request.xhr?
-        flash[:success] = 'Logged In'
-        redirect redirect_path
+        if multi_factor_authentication?
+          Services::Authentication.prep_mfa(current_user_id, request)
+          flash[:info] = 'An email was sent with your OTP'
+          mfa_url = "#{settings.map_path}/auth/identity/mfa"
+          redirect mfa_url
+        else
+          successful_login
+        end
       else
-        # Failed Login
-        broadcast(:identity_failed_login, target: self, details: "IP: #{request.ip}")
-        flash[:warning] = 'Invalid credentials. Please try again.'
+        failed_login
+      end
+    end
+
+    get '/auth/identity/mfa' do
+      if current_user&.is_a? Ditty::User
+        identity = Identity.find(user_id: current_user_id)
+        unless identity.pin_expired?
+          return haml :'identity/mfa', locals: { identity: identity, title: 'Enter OTP' }
+        end
+
+        logout
         redirect "#{settings.map_path}/auth/identity"
       end
     end
 
+    post '/auth/identity/mfa' do
+      if current_user&.is_a? Ditty::User
+        halt 500 unless params[:otp]
```

Comment:
Use Sinatra Param to check this. It should return a 400, not a 500 (which sinatra param will handle).